README.md
# ui-progresspie


An Angular directive which renders a circular progress meter with two values: Expected and Actual progress (0.0 - 1.0)


## Usage

Include the `ui-progresspie` dependency on your Angular module:
```
var app = angular.module('demoapp', ['ui-progresspie']);
```


Then embed the progress meter in your page:

```html
<div progresspie actual="0.5" expected="0.1" size="240" threshold="0.1"></div>
```

#TODO:

* Ensure multiple can be on the same page
* Write unit tests
* What if it's placed on a non-white document?
* Think about backwards compatibility (Sizzle?)  
* Coffee lint
* Travis?
* Error handling - invalid values currently break it
* More tests!


CHANGELOG.md
<a name"0.0.1"></a>
### 0.0.1 (2016-02-11)

#### Features

* **Initial Commit:** First version. Not quite working.


test/karma.conf.js
// Karma configuration
// Generated on Tue Feb 16 2016 16:48:02 GMT-0800 (PST)

module.exports = function(config) {
  config.set({

    // base path that will be used to resolve all patterns (eg. files, exclude)
    basePath: '../',


    // frameworks to use
    // available frameworks: https://npmjs.org/browse/keyword/karma-adapter
    frameworks: ['jasmine'],


    // list of files / patterns to load in the browser
    files: [
      // deps
      'node_modules/jquery/dist/jquery.js',
      'node_modules/angular/angular.js',
      'node_modules/angular-mocks/angular-mocks.js',
      'node_modules/d3/d3.js',
      // src
      'dist/progresspie.js',
      // tests
      'test/spec/directive/progresspie-spec.js'
    ],


    // preprocess matching files before serving them to the browser
    // available preprocessors: https://npmjs.org/browse/keyword/karma-preprocessor
    preprocessors: {
    },


    // test results reporter to use
    // possible values: 'dots', 'progress'
    // available reporters: https://npmjs.org/browse/keyword/karma-reporter
    reporters: ['progress'],


    // web server port
    port: 9876,


    // enable / disable colors in the output (reporters and logs)
    colors: true,


    // level of logging
    // possible values: config.LOG_DISABLE || config.LOG_ERROR || config.LOG_WARN || config.LOG_INFO || config.LOG_DEBUG
    logLevel: config.LOG_INFO,


    // enable / disable watching file and executing tests whenever any file changes
    autoWatch: true,


    // start these browsers
    // available browser launchers: https://npmjs.org/browse/keyword/karma-launcher
    browsers: ['Chrome'],


    // Continuous Integration mode
    // if true, Karma captures browsers, runs the tests and exits
    singleRun: false,

    // Concurrency level
    // how many browser should be started simultaneous
    concurrency: Infinity
  })
}


test/spec/directive/progresspie-spec.js
describe('progress-pie', function() {
  var elm, scope;

  beforeEach(module('ui-progresspie'));

  beforeEach(inject(function($rootScope, $compile) {
    elm = angular.element('<div progresspie expected="0" actual="0" threshold="0.1" size="240"></div>');
    scope = $rootScope;
    $compile(elm)(scope);
    scope.$digest();
  }));

  it('creates an svg node', inject(function($compile, $rootScope) {
    expect($(elm).find('svg').length).toEqual(1);
  }));

  it('defaults to a 240 pixel square', inject(function($compile, $rootScope) {
    expect($(elm).find('svg').attr('width')).toBe('240');
    expect($(elm).find('svg').attr('height')).toBe('240');
  }));

  it('displays a 0 in progress', inject(function($compile, $rootScope) {
    expect($(elm).find('text tspan:first').text()).toBe('0');
  }));

  it('updates actual progress number text', inject(function($compile, $rootScope) {
    scope.ctrl.actual = 0.5;
    scope.ctrl.update();
    expect($(elm).find('text tspan:first').text()).toBe('50');
  }));

  it('becomes in danger when it falls outside threshold', inject(function($compile, $rootScope) {
    scope.ctrl.actual = 0.1;
    scope.ctrl.threshold = 0.5;
    scope.ctrl.update();
    //TODO: Determine danger
  }));
});


test/spec/directives/progresspie-spec.js
describe('tabs', function() {
  var elm, scope;

  // load the tabs code
  beforeEach(module('ui-progresspie'));

  // load the templates
  // beforeEach(module('tpl/tabs.html', 'tpl/pane.html'));

  beforeEach(inject(function($rootScope, $compile) {
    elm = angular.element('<div progresspie></div>');
    scope = $rootScope;
    $compile(elm)(scope);
    scope.$digest();
  }));


  it('should create clickable titles', inject(function($compile, $rootScope) {
    // var titles = elm.find('ul.nav-tabs li a');

    // expect(titles.length).toBe(2);
    // expect(titles.eq(0).text()).toBe('First Tab');
    // expect(titles.eq(1).text()).toBe('Second Tab');
  }));
});


js/progresspie.js

/**
  * @ngdoc directive
  * @name ui-progresspie.directive:progresspie
  * @description
  * # progressPie
 */

(function() {
  var ProgressPie;

  angular.module('ui-progresspie', []).directive('progresspie', function() {
    return {
      resrict: 'EA',
      controller: 'ProgressPieCtrl as ctrl',
      link: function(scope, element, attr, ctrl) {
        scope.$watch('ctrl.actual', function() {
          return ctrl.update();
        });
        scope.$watch('ctrl.expected', function() {
          return ctrl.update();
        });
        scope.$watch('ctrl.threshold', function() {
          return ctrl.update();
        });
        scope.$watch('ctrl.size', function() {
          return ctrl.update();
        });
        return ctrl.draw(element);
      },
      bindToController: {
        actual: '=',
        expected: '=',
        threshold: '=',
        size: '='
      }
    };
  }).controller('ProgressPieCtrl', ProgressPie = (function() {
    ProgressPie.prototype.actual = 0;

    ProgressPie.prototype.expected = 0;

    ProgressPie.prototype.threshold = 0.1;

    ProgressPie.prototype.normalColor = "#78c000";

    ProgressPie.prototype.dangerColor = "#f00";

    ProgressPie.prototype.size = 240;

    ProgressPie.prototype.updateSpeed = 500;

    function ProgressPie() {
      this.radius = this.size / 2;
      this.tau = 2 * Math.PI;
    }

    ProgressPie.prototype.draw = function(element) {
      var actualInner, actualOuter, centerCircle, circleRadius, danger, expectedInner, expectedOuter, fontSize, gap, lineGap, meters, progressLabel, progressVal, radius, svg, textGroup;
      this.elem = d3.select(element[0]);
      radius = this.size / 2;
      actualInner = this.radius - (this.radius * 0.10);
      actualOuter = this.radius;
      gap = this.radius * 0.02;
      expectedInner = this.radius - (this.radius * 0.17);
      expectedOuter = actualInner - gap;
      circleRadius = this.radius * 0.7;
      danger = false;
      fontSize = this.radius * 0.4;
      lineGap = fontSize * 0.65;
      this.actualArc = d3.svg.arc().innerRadius(actualInner).outerRadius(actualOuter).cornerRadius(90).startAngle(0);
      this.expectedArc = d3.svg.arc().innerRadius(expectedInner).outerRadius(expectedOuter).cornerRadius(90).startAngle(0);
      svg = this.elem.append("svg").attr("width", this.size).attr("height", this.size).append("g").attr("transform", "translate(" + this.radius + "," + this.radius + ")");
      centerCircle = svg.append("circle").attr("cx", 0).attr("cy", 0).attr("r", circleRadius).style("fill", "#000").style("opacity", "0.05");
      textGroup = svg.append("g").attr("alignment-baseline", "middle");
      progressVal = textGroup.append("text").attr("text-anchor", "middle");
      this.progressNum = progressVal.append("tspan").text("0").attr("font-size", fontSize);
      progressVal.append("tspan").attr("font-size", fontSize / 2).text("%");
      progressLabel = textGroup.append("text").attr("font-size", fontSize / 2).attr("transform", "translate(0, " + lineGap + ")").attr("text-anchor", "middle").text("Progress");
      meters = svg.append("g").attr("fill", this.normalColor);
      this.actualPath = meters.append("path").datum({
        endAngle: 0
      }).attr("d", this.actualArc);
      return this.expectedPath = meters.append("path").datum({
        endAngle: 0
      }).attr("d", this.expectedArc).attr("opacity", "0.33");
    };

    ProgressPie.prototype.isDanger = function() {
      return this.expected - this.actual >= this.threshold;
    };

    ProgressPie.prototype.update = function() {
      var endColor;
      this.progressNum.text(Math.round(this.actual * 100));
      endColor = this.isDanger() ? this.dangerColor : this.normalColor;
      console.log(endColor);
      this.actualPath.transition().duration(this.updateSpeed).style("fill", endColor).call(this.arcTween, this.actual * this.tau, this.actualArc);
      return this.expectedPath.transition().duration(this.updateSpeed).style("fill", endColor).call(this.arcTween, this.expected * this.tau, this.expectedArc);
    };

    ProgressPie.prototype.arcTween = function(transition, newAngle, arc) {
      return transition.attrTween("d", function(d) {
        var interpolate;
        interpolate = d3.interpolate(d.endAngle, newAngle);
        return function(t) {
          d.endAngle = interpolate(t);
          return arc(d);
        };
      });
    };

    ProgressPie.prototype.setRandom = function() {
      return this.update(Math.random(), Math.random());
    };

    return ProgressPie;

  })());

}).call(this);
